samples/subsys/usb/hid-mouse/src/main.c

```diff
@@ -1,245 +1,117 @@
-	}
 
-	if (status[MOUSE_BTN_REPORT_POS] != state) {
-		status[MOUSE_BTN_REPORT_POS] = state;
-		k_sem_give(&sem);
-	}
-}
+		break;
+	case INPUT_KEY_1:
+		rwup_if_suspended();
+		idx = MOUSE_BTN_REPORT_IDX;
+		value = report[idx];
 
-static void x_move(const struct device *gpio, struct gpio_callback *cb,
-		   uint32_t pins)
-{
-	int ret;
-	uint8_t state = status[MOUSE_X_REPORT_POS];
+		if (evt->value) {
+			value |= MOUSE_BTN_RIGHT;
+		} else {
+			value &= ~MOUSE_BTN_RIGHT;
+		}
 
-	ret = gpio_pin_get(gpio, sw2.pin);
-	if (ret < 0) {
-		LOG_ERR("Failed to get the state of port %s pin %u, error: %d",
-			gpio->name, sw2.pin, ret);
-		return;
-	}
+		break;
+	case INPUT_KEY_2:
+		idx = MOUSE_X_REPORT_IDX;
+		value = report[idx];
 
-	if (def_val[2] != (uint8_t)ret) {
-		state += 10U;
-	}
+		if (evt->value) {
+			value += 10U;
+		}
 
-	if (status[MOUSE_X_REPORT_POS] != state) {
-		status[MOUSE_X_REPORT_POS] = state;
-		k_sem_give(&sem);
-	}
-}
+		break;
+	case INPUT_KEY_3:
+		idx = MOUSE_Y_REPORT_IDX;
+		value = report[idx];
 
-static void y_move(const struct device *gpio, struct gpio_callback *cb,
-		   uint32_t pins)
-{
-	int ret;
-	uint8_t state = status[MOUSE_Y_REPORT_POS];
+		if (evt->value) {
+			value += 10U;
+		}
 
-	ret = gpio_pin_get(gpio, sw3.pin);
-	if (ret < 0) {
-		LOG_ERR("Failed to get the state of port %s pin %u, error: %d",
-			gpio->name, sw3.pin, ret);
+		break;
+	default:
 		return;
 	}
 
-	if (def_val[3] != (uint8_t)ret) {
-		state += 10U;
-	}
-
-	if (status[MOUSE_Y_REPORT_POS] != state) {
-		status[MOUSE_Y_REPORT_POS] = state;
-		k_sem_give(&sem);
+	if (report[idx] != value) {
+		report[idx] = value;
+		k_sem_give(&report_sem);
 	}
 }
 
-int callbacks_configure(const struct gpio_dt_spec *spec,
-			gpio_callback_handler_t handler,
-			struct gpio_callback *callback, uint8_t *val)
-{
-	const struct device *gpio = spec->port;
-	gpio_pin_t pin = spec->pin;
-	int ret;
-
-	if (gpio == NULL) {
-		/* Optional GPIO is missing. */
-		return 0;
-	}
-
-	if (!device_is_ready(gpio)) {
-		LOG_ERR("GPIO port %s is not ready", gpio->name);
-		return -ENODEV;
-	}
-
-	ret = gpio_pin_configure_dt(spec, GPIO_INPUT);
-	if (ret < 0) {
-		LOG_ERR("Failed to configure port %s pin %u, error: %d",
-			gpio->name, pin, ret);
-		return ret;
-	}
-
-	ret = gpio_pin_get(gpio, pin);
-	if (ret < 0) {
-		LOG_ERR("Failed to get the state of port %s pin %u, error: %d",
-			gpio->name, pin, ret);
-		return ret;
-	}
-
-	*val = (uint8_t)ret;
-
-	gpio_init_callback(callback, handler, BIT(pin));
-	ret = gpio_add_callback(gpio, callback);
-	if (ret < 0) {
-		LOG_ERR("Failed to add the callback for port %s pin %u, "
-			"error: %d",
-			gpio->name, pin, ret);
-		return ret;
-	}
-
-	ret = gpio_pin_interrupt_configure_dt(spec, GPIO_INT_EDGE_BOTH);
-	if (ret < 0) {
-		LOG_ERR("Failed to configure interrupt for port %s pin %u, "
-			"error: %d",
-			gpio->name, pin, ret);
-		return ret;
-	}
-
-	return 0;
-}
+INPUT_LISTENER_CB_DEFINE(NULL, input_cb);
```

Comment:
Do we want this to get input for any potential device? Guess it's fine for the sample, otherwise you could do a get any gpio-keys here.

Reply:
Yes